src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,137 @@
+        } else if (index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index is too much");
+        } else if (index == size) {
+            add(value);
+        } else {
+            if (values.length == size) {
+                grow();
+            }
+            T[] newValues = (T[]) new Object[values.length];
+            for (int i = 0; i < index; i++) {
+                newValues[i] = values[i];
+            }
+            newValues[index] = value;
+            for (int i = index; i < size; i++) {
+                newValues[i + 1] = values[i];
+            }
+            size++;
+            values = newValues;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("No such element exists");
+        } else if (index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't get element by negative index");
+        }
+        return values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("No such index exists");
+        } else if (index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't set to negative position");
+        }
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Can't remove by negative index");
+        } else if (index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("No such element exists");
+        }
+        T[] newValues = (T[]) new Object[values.length];
+        for (int i = 0; i < index; i++) {
+            newValues[i] = values[i];
+        }
+        for (int i = index + 1; i < values.length; i++) {
+            newValues[i - 1] = values[i];
+        }
+        T result = values[index];
+        values = newValues;
+        size--;
+        return result;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T[] newValues = (T[]) new Object[values.length];
+        int elementIndex = 0;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(values[i], element)) {
+                elementIndex = i;
+                break;
+            }
+            if (i == size - 1) {
+                throw new NoSuchElementException("No such element exists");
+            }
+        }
+        for (int i = 0; i < elementIndex; i++) {
+            newValues[i] = values[i];
+        }
+        for (int i = elementIndex + 1; i < size; i++) {
+            newValues[i - 1] = values[i];
+        }
+        values = newValues;
+        size--;
+        return element;
```

Comment:
There are too much extra lines))))
```suggestion
        for (int i = 0; i < size; i++) {
            if (element == values[i] || (values[i] != null && values[i].equals(element))) {
                return remove(i);
            }
        }
        throw new NoSuchElementException("There is no such element: " + element);
```
It wil be enough)